### PR TITLE
fix(windows): make keymanx64 responsible for its own lifecycle

### LINKED ISSUE #4976: chore(windows): change responsibility for keymanx64 shutdown
**Is your feature request related to a problem? Please describe.**

Right now, keymanx64 shuts down when it is posted a `WM_USER_PlatformComm` + `PC_CLOSE` message. However, it would be much better if its lifetime was determined by watching for the shutdown of its parent keyman.exe. The best way to do this is probably via `WaitForSingleObject` in a thread that then posts this same message to the main window.

Moving the responsibility for the lifetime of keymanx64 into keymanx64 makes it easier to reason about the keymanx64 state machine. It deals with situations where keyman.exe is terminated abnormally. It also resolves issues around multiple instances of keymanx64, because we don't allow multiple instances of keyman.exe (keyman.exe detects if it is already running and shuts down if so), and so any time keyman.exe shuts down we'll cleanup pretty much instantly.

This may remove the need for a mutex for keymanx64.exe: it should be passed a process handle for its parent keyman.exe, and if that is not present, then it should fail. Furthermore, keyman.exe could watch for the abnormal termination of keymanx64.exe and restart it. (Note: passing a process handle avoids a slight possibility of a race condition -- the handle will be to the parent keyman even if it exits during process startup of keymanx64. However, passing a process id is likely to work because Windows does not reuse process IDs rapidly; it is just slightly less correct.)

I believe that this would avoid the majority of keymanx64 startup errors that we currently see reported in Sentry.

---

**Keyman for Windows/macOS/Linux:**
 - OS: Windows 10 
 - Keyman version: 14.0.272

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #4976.
Fixes KEYMAN-WINDOWS-2K.
Fixes KEYMAN-WINDOWS-3F.
Fixes KEYMAN-WINDOWS-5A.
Fixes KEYMAN-WINDOWS-3E.

This fix reworks keymanx64's lifecycle, moving responsibility for process shutdown from keyman.exe to keymanx64.exe.

This eliminates the need for interprocess communication, and simplifies the startup and shutdown of keymanx64.exe. Removing this means that we can more safely handle situations where two instances of keymanx64 may be started, as one of them will rapidly terminate when it discovers that its parent process has disappeared.

I will cherry-pick this to stable-14.0 once we have run this on alpha 15 for a short while. It is a bit of a change in design and while I believe the changes are robust, better safe than sorry.

### CODE REVIEW

**Comment by mcdurdin** [comment]
One question: is there any chance that Keyman32 would ever have its main window recreated? (This is possible in Delphi's form model which recreates windows in certain circumstances. See [RecreateWnd](http://docwiki.embarcadero.com/Libraries/Sydney/en/Vcl.Controls.TWinControl.RecreateWnd), [SO](https://stackoverflow.com/questions/27400973/what-external-events-trigger-tcustomform-recreatewnd)) Doing so would invalidate the handle passed into keymanx64 as a commandline parameter and cause it to work incorrectly. AFAIK, `Application.Handle` is never recreated.

I will do some investigation on this before merging this PR. But it is still ready for review apart from this.

**Comment by mcdurdin** [comment]
I've split that question above into a separate issue at #5004. It need not hold up this PR, because it's a longstanding issue if it exists.

File: windows/src/engine/keymanx64/keymanx64.cpp
Comment (darcywong00) [inline windows/src/engine/keymanx64/keymanx64.cpp:391]: Does cleanup work when either `hWatcherThreadTerminateEvent` or `hWatcherThread` are NULL?
(for when `CreateWatcherThread()` fails)

Comment (mcdurdin) [reply windows/src/engine/keymanx64/keymanx64.cpp:391]: Good question. Turns out `WM_DESTROY` is called even if `WM_CREATE` returns `FALSE`, so it is important to cleanup correctly. Done now!

**Comment by mcdurdin** [comment]
Apologies for the force-push, committed the wrong files by mistake. Should be clean again now.

**Review by rc-swag — APPROVED** [review]
lgtm

**Review by darcywong00 — APPROVED** [review]
lgtm

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.43-alpha](https://keyman.com/downloads/releases/alpha/15.0.43)

**Pull request merged**
